### DISCUSSION TITLE
Any plans on having mobile versions of ecode on android and ios?

### DISCUSSION THREAD

**Discussion by emeraldtip** [SpartanJ/ecode#92] [Ideas]
I know quite a few people who'd find that pretty useful. Even without the terminal.
Imagine participating in codeforces competitions on your tablet

**Comment by SpartanJ** [comment]
It's actually possible, I just don't build it because it's un-usable in a small screen. But I even added terminal support. I can build you an Android apk and give me your opinion. I'll later update the build script an uploaded here.

**Reply by SpartanJ** [reply]
Here's an android build: [ecode-android-0.4.8.zip](https://github.com/SpartanJ/ecode/files/11130908/ecode-android-0.4.8.zip)

**Reply by emeraldtip** [reply]
Thanks!

I played around with it a bit
Firstly yeah, it's fairly unusable on a phone screen as the tiny settings icon is right in the corner of the screen and I can't touch it. (works with a mouse tho)
Also some of the menus go off-screen:
![Screenshot_20230403_084621_ecode](https://user-images.githubusercontent.com/48155462/229423973-9ea57294-cf97-4a76-bc4b-3828e8b7f0f8.png)

There also appears to be an issue on start up where the screen is entirely blank until you click on something.

I also have a suggestion - could the Termux shell be implemented? With the normal android shell it's just impossible to run for example python scripts (there's probably some method if you root the device, but that's fairly annoying to do)

**Reply by SpartanJ** [reply]
> Firstly yeah, it's fairly unusable on a phone screen as the tiny settings icon is right in the corner of the screen and I can't touch it. (works with a mouse tho)

Yeah, I don't think it's fixable with a few tweaks, you can already change the UI scale factor but won't do magic.

> Also some of the menus go off-screen:

Yes and it can't be fixed using these kinds of menus, they should be different for a mobile version.

> There also appears to be an issue on start up where the screen is entirely blank until you click on something.

True, I didn't see it the first time, it's trivial to fix it though.

> I also have a suggestion - could the Termux shell be implemented? With the normal android shell it's just impossible to run for example python scripts (there's probably some method if you root the device, but that's fairly annoying to do)

It's not possible in Android. We don't have normal binaries for the applications, so I can't consume them, mobile OSes are crap. We should never use a mobile OS for development. To have a real shell like in Termux it would be needed to somehow provide the shell with the app, that's what Termux does (Termux is just a terminal emulator with very good tooling).

I know that with some work it could be possible to have a decent ecode version on Android, but I don't see the point, it would move the development focus to something that it's really not meaningful for the moment. These are the kind of sub-projects that would be cool for a collaborator, but I won't use any precious time in improving the Android experience for the moment, there's so much to do yet that it doesn't make sense. If you ask me it's more interesting as a sub-project the [web version](https://cdn.ensoft.dev/eepp-demos/demo-fs.html?run=ecode.js).

**Comment by ghost** [comment]
The issue with the menu goes off-screen is easy to fix. Have a look at how applications like Firefox and Chrome do it on Android. You simply have to set it to discard the parent menu to have space for the sub-menu and done.

This is what applications for vertical screens should do. ecode is too tailored towards wide horizontal screens (of the desktop). BTW, there are people use vertical screens even on the desktop. You rotate the screen and set the desktop environment to adjust the screen orientation.

**Reply by ghost** [reply]
On touch screens, the users have to tap to open the sub-menu instead of having it to spawn automatically with mouse hovering like on desktop. You can't expect that there is always a mouse available.

The menu of ecode reminds me of the right click menu of Openbox/Fluxbox. The sub-menus will spawn indefinitely.
